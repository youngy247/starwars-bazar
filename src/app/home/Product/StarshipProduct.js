import useAddToBasket from '@/hooks/useAddToBasket';
import { Add, Subtract } from '@carbon/icons-react';
import { Button, ToastNotification } from '@carbon/react';
import React, { useState } from 'react';

export default function StarshipProduct({ starship }) {
  const [quantity, setQuantity] = useState(1);
  const { toastVisible, toastMessage, addToBasket } = useAddToBasket();

  const handleAddToBasket = () => {
    addToBasket(quantity, starship.name);
  };

  return (
    <section className={'starship'}>
      <h1>{starship.name}</h1>
      <p>
        <strong>Hyperdrive Rating:</strong> {starship.hyperdrive_rating}
      </p>
      <p>
        <strong>Passengers:</strong> {starship.passengers}
      </p>
      <p>
        <strong>Cost In Credits:</strong> {starship.cost_in_credits}
      </p>
      <p>
        <strong>Manufacturer:</strong> {starship.manufacturer}
      </p>
      <p>
        <strong>Model:</strong> {starship.model}
      </p>
      <p>
        <strong>Created At:</strong>{' '}
        {new Date(starship.created).toLocaleDateString()}
      </p>

      <div className={'quantity-container'}>
        <div className={'quantity-value-container'}>
          <Button
            renderIcon={Subtract}
            hasIconOnly
            onClick={() => setQuantity(quantity - 1)}
            disabled={quantity <= 1}
            size="sm"
            kind="secondary"
            title={`Decrease quantity of ${starship.name} to quantity of ${
              quantity - 1
            }`}
            iconDescription={`Decrease quantity of ${
              starship.name
            } to quantity of ${quantity - 1}`}></Button>
          <span>{quantity}</span>
          <Button
            onClick={() => setQuantity(quantity + 1)}
            hasIconOnly
            renderIcon={Add}
            size="sm"
            kind="secondary"
            title={`Increase quantity of ${starship.name} to quantity of ${
              quantity + 1
            }`}
            iconDescription={`Increase quantity of ${
              starship.name
            } to quantity of ${quantity + 1}`}
          />
        </div>
        <Button onClick={handleAddToBasket} size="lg" kind="primary">
          BUY
        </Button>
      </div>

      {toastVisible && (
        <ToastNotification
          role="status"
          kind="success"
          timeout={3000}
          subtitle={toastMessage}
          title="Success"
          caption="View your basket"
          className={'toast-notification'}
        />
      )}
    </section>
  );
}
