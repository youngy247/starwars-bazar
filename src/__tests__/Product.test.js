import StarshipProduct from '@/app/home/Product/StarshipProduct';
import '@testing-library/jest-dom';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';

const mockStarship = {
  name: 'Jedi Interceptor',
  model: 'Eta-2 Actis-class light interceptor',
  manufacturer: 'Kuat Systems Engineering',
  cost_in_credits: '320000',
  length: '5.47',
  max_atmosphering_speed: '1500',
  crew: '1',
  passengers: '0',
  cargo_capacity: '60',
  consumables: '2 days',
  hyperdrive_rating: '1.0',
  MGLT: 'unknown',
  starship_class: 'starfighter',
  pilots: [
    'https://swapi.dev/api/people/10/',
    'https://swapi.dev/api/people/11/',
  ],
  films: ['https://swapi.dev/api/films/6/'],
  created: '2014-12-20T19:56:57.468000Z',
  edited: '2014-12-20T21:23:49.951000Z',
  url: 'https://swapi.dev/api/starships/65/',
};

describe('Starship Component', () => {
  test('renders starship details correctly', () => {
    render(<StarshipProduct starship={mockStarship} />);

    expect(screen.getByText('Jedi Interceptor')).toBeInTheDocument();
    expect(screen.getByText('Hyperdrive Rating:')).toBeInTheDocument();
    expect(screen.getByText('1.0')).toBeInTheDocument();
    expect(screen.getByText('Passengers:')).toBeInTheDocument();
    expect(screen.getByText('0')).toBeInTheDocument();
    expect(screen.getByText('Cost In Credits:')).toBeInTheDocument();
    expect(screen.getByText('320000')).toBeInTheDocument();
    expect(screen.getByText('Manufacturer:')).toBeInTheDocument();
    expect(screen.getByText('Kuat Systems Engineering')).toBeInTheDocument();
  });

  test('handles quantity increment and decrement', () => {
    render(<StarshipProduct starship={mockStarship} />);

    const minusButton = screen.getByRole('button', {
      name: /Decrease quantity of Jedi Interceptor to quantity of \d+/i,
    });
    const plusButton = screen.getByRole('button', {
      name: /Increase quantity of Jedi Interceptor to quantity of \d+/i,
    });

    const quantityDisplay = screen.getByText('1');

    expect(quantityDisplay).toHaveTextContent('1');

    act(() => {
      fireEvent.click(plusButton);
    });

    expect(screen.getByText('2')).toBeInTheDocument();

    act(() => {
      fireEvent.click(minusButton);
    });

    expect(screen.getByText('1')).toBeInTheDocument();

    act(() => {
      fireEvent.click(minusButton);
    });

    expect(minusButton).toBeDisabled();
  });

  test('displays toast notification when adding to basket', async () => {
    render(<StarshipProduct starship={mockStarship} />);

    const buyButton = screen.getByText('BUY');

    act(() => {
      fireEvent.click(buyButton);
    });

    await waitFor(() => {
      expect(screen.getByText('Success')).toBeInTheDocument();
    });
  });

  test('hides toast notification after timeout', async () => {
    jest.useFakeTimers();
    render(<StarshipProduct starship={mockStarship} />);

    act(() => {
      fireEvent.click(screen.getByText('BUY'));
    });

    expect(screen.getByText('Success')).toBeInTheDocument();

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    await waitFor(() => {
      expect(screen.queryByText('Success')).not.toBeInTheDocument();
    });

    jest.useRealTimers();
  });
});
