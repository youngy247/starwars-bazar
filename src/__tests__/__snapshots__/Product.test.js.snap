// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Starship Component matches snapshot 1`] = `
<section
  className="starship"
>
  <h1>
    Jedi Interceptor
  </h1>
  <p>
    <strong>
      Hyperdrive Rating:
    </strong>
     
    1.0
  </p>
  <p>
    <strong>
      Passengers:
    </strong>
     
    0
  </p>
  <p>
    <strong>
      Cost In Credits:
    </strong>
     
    320000
  </p>
  <p>
    <strong>
      Manufacturer:
    </strong>
     
    Kuat Systems Engineering
  </p>
  <p>
    <strong>
      Model:
    </strong>
     
    Eta-2 Actis-class light interceptor
  </p>
  <p>
    <strong>
      Created At:
    </strong>
     
    20/12/2014
  </p>
  <div
    className="quantity-container"
  >
    <div
      className="quantity-value-container"
    >
      <span
        className="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top cds--tooltip cds--icon-tooltip cds--icon-tooltip--disabled"
        onKeyDown={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="cds--tooltip-trigger__wrapper"
        >
          <button
            aria-labelledby="tooltip-:r0:"
            className="cds--btn--icon-only cds--btn cds--btn--sm cds--layout--size-sm cds--btn--secondary cds--btn--disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            onTouchStart={[Function]}
            title="Decrease quantity of Jedi Interceptor to quantity of 0"
            type="button"
          >
            <svg
              aria-hidden={true}
              fill="currentColor"
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 15H24V17H8z"
              />
            </svg>
          </button>
        </div>
        <span
          aria-hidden="true"
          className="cds--popover"
          id="tooltip-:r0:"
          onMouseEnter={[Function]}
          role="tooltip"
        >
          <span
            className="cds--popover-content cds--tooltip-content"
          >
            Decrease quantity of Jedi Interceptor to quantity of 0
          </span>
          <span
            className="cds--popover-caret"
          />
        </span>
      </span>
      <span>
        1
      </span>
      <span
        className="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top cds--tooltip cds--icon-tooltip"
        onKeyDown={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="cds--tooltip-trigger__wrapper"
        >
          <button
            aria-labelledby="tooltip-:r3:"
            className="cds--btn--icon-only cds--btn cds--btn--sm cds--layout--size-sm cds--btn--secondary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            onTouchStart={[Function]}
            title="Increase quantity of Jedi Interceptor to quantity of 2"
            type="button"
          >
            <svg
              aria-hidden={true}
              fill="currentColor"
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
              />
            </svg>
          </button>
        </div>
        <span
          aria-hidden="true"
          className="cds--popover"
          id="tooltip-:r3:"
          onMouseEnter={[Function]}
          role="tooltip"
        >
          <span
            className="cds--popover-content cds--tooltip-content"
          >
            Increase quantity of Jedi Interceptor to quantity of 2
          </span>
          <span
            className="cds--popover-caret"
          />
        </span>
      </span>
    </div>
    <button
      className="cds--btn cds--layout--size-lg cds--btn--primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      BUY
    </button>
  </div>
</section>
`;
